### PR TITLE
Alerting: Hook up GMA silence APIs to new authentication handler

### PR DESCRIPTION
**Pull request by JacobsonMT**
**What is this feature?**

This PR connects the new RBAC authentication service to existing alertmanager API silence endpoints.

**Why do we need this feature?**

This PR is an incremental improvement towards fine-grained RBAC for alertmanager configurations, specifically for silences.

#### What functionality does this change allow?

**OSS**
This allows giving limited-access users/teams/roles/services accounts elevated access to view/create/edit silences for alerts in specific folders. This is done through the manage permissions page for folders.

View: Access to view rule-specific silences in the given folder.
Edit/Admin: Access to view, create, and update rule-specific silences in the given folder.

**Enterprise**
In addition to the above OSS functionality, the new permissions will be in effect when used in modified built-in roles, custom fixed roles or other aspects of enterprise RBAC.

See https://github.com/grafana/grafana/pull/85418 for more details on the specific permissions.

#### Backwards compatibility

**OSS**
`View/Edit/Admin` permissions on a folder give the equivalent silence permissions for rules in that folder.

**Enterprise**
In addition to the above OSS backwards compatibility changes:

- Users with Role `None` that also have any of `View/Edit/Admin` on a folder will also have access to view, but not modify/create, silences not attached specifically to a single rule (general silences vs rule-specific),
- `alert.instances.read` is required in addition to `alert.instances.create/write` to perform the create/write CRUD operations. This should have no effect on OSS users as the permissions always come together. Enterprise custom roles that contained `alert.instances.create/write` without `alert.instances.read` will need to amend the role definition.
- Fixed roles "Rules Reader"(`fixed:alerting.rules:read`) and  "Rules Writer"(`fixed:alerting.rules:writer`) are by default given silence read and silence read+write+create permissions, respectively. This means, for example, that users with "Rules Reader" will now see all silences where they previously would not have and users with "Rules Reader" will now see all silences and have write/create permission for rule-specific silences.

#### UI/UX impact

It should have very little, if any, effect in the UI until we implement UI support for the new `alert.silences.read/write/update` permissions. This change currently applies at the API level.

#### Notes
- The bulk of the PR is for testing utilities + integration tests for the existing silence endpoints.
- We could consider not adding the ngalert model `Silence` and instead implementing `GetRuleUID` in `grafana/alerting`.
- Added an extra `alertmanagerForOrg` method to `MultiOrgAlertmanager`. This will go away eventually once our API layer no longer access the bare alertmanagers directly. The changes necessary to do this now are out of scope of this PR and not small/related to enough to squeeze it in.

### CODE REVIEW

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go [hunk 1]
```diff
@@ -400,27 +413,90 @@ func (moa *MultiOrgAlertmanager) AlertmanagerFor(orgID int64) (Alertmanager, err
...
+	moa.alertmanagersMtx.RLock()
+	defer moa.alertmanagersMtx.RUnlock()
+
+	orgAM, err := moa.alertmanagerForOrg(orgID)
 	if err != nil {
 		return "", err
 	}
 
+	// Need to create the silence in the AM first to get the silence ID.
+	silenceID, err := orgAM.CreateSilence(ctx, SilenceToNotifyPostableSilence(ps))
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/notifier/multiorg_alertmanager.go:488]: I wonder if we should rename it to CreateOrUpdateSilence

File: pkg/services/ngalert/notifier/silence_svc.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+type SilenceAccessControlService interface {
+	FilterByAccess(ctx context.Context, user identity.Requester, silences ...*models.Silence) ([]*models.Silence, error)
+	AuthorizeReadSilence(ctx context.Context, user identity.Requester, silence *models.Silence) error
+	AuthorizeCreateSilence(ctx context.Context, user identity.Requester, silence *models.Silence) error
+	AuthorizeUpdateSilence(ctx context.Context, user identity.Requester, silence *models.Silence) error
+}
+
+// SilenceStateStore is the interface for storing and retrieving silences. Currently, this is implemented by
+// MultiOrgAlertmanager but should eventually be replaced with an actual store.
+type SilenceStateStore interface {
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/notifier/silence_svc.go:29]: why do you use "state"? I think just silence store is clear enough

File: pkg/tests/api/alerting/api_alertmanager_silence_test.go [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+		orgRole          org.RoleType // default RoleNone
+		permissions      []accesscontrol.Permission
+		defaultAllowed   bool                                  // Default allowed/forbidden for actions not in statusExceptions.
+		statusExceptions map[silenceType]map[silenceAction]int // Exceptions to defaultAllowed.
+		listContents     []silenceType                         // nil = forbidden.
+	}{
+		// OSS Builtins
+		{
+			name:    "No permissions",
+			orgRole: org.RoleNone,
```
Comment (yuri-tceretian) [inline pkg/tests/api/alerting/api_alertmanager_silence_test.go:108]: This role is not permitted in OSS

File: pkg/services/ngalert/api/authorization.go [hunk 4]
```diff
@@ -119,16 +119,17 @@ func (api *API) authorize(method, path string) web.Handler {
...
 	case http.MethodGet + "/api/alertmanager/grafana/api/v2/silence/{SilenceId}":
-		eval = ac.EvalPermission(ac.ActionAlertingInstanceRead)
+		eval = ac.EvalAny(ac.EvalPermission(ac.ActionAlertingInstanceRead), ac.EvalPermission(ac.ActionAlertingSilencesRead))
 	case http.MethodGet + "/api/alertmanager/grafana/api/v2/silences":
-		eval = ac.EvalPermission(ac.ActionAlertingInstanceRead)
+		eval = ac.EvalAny(ac.EvalPermission(ac.ActionAlertingInstanceRead), ac.EvalPermission(ac.ActionAlertingSilencesRead))
 	case http.MethodPost + "/api/alertmanager/grafana/api/v2/silences":
 		// additional authorization is done in the request handler
-		eval = ac.EvalAny(ac.EvalPermission(ac.ActionAlertingInstanceCreate), ac.EvalPermission(ac.ActionAlertingInstanceUpdate))
+		eval = ac.EvalAny(ac.EvalPermission(ac.ActionAlertingInstanceCreate), ac.EvalPermission(ac.ActionAlertingInstanceUpdate), ac.EvalPermission(ac.ActionAlertingSilencesCreate), ac.EvalPermission(ac.ActionAlertingSilencesWrite))
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/api/authorization.go:132]: Let's add read permissions here too

File: pkg/services/ngalert/api/authorization.go [hunk 5]
```diff
@@ -119,16 +119,17 @@ func (api *API) authorize(method, path string) web.Handler {
 
 	// Alert Instances and Silences
 
-	// Silences. Grafana Paths
+	// Silences. Grafana Paths.
+	// These permissions are required but not sufficient, further authorization is done in the request handler.
 	case http.MethodDelete + "/api/alertmanager/grafana/api/v2/silence/{SilenceId}":
-		eval = ac.EvalPermission(ac.ActionAlertingInstanceUpdate) // delete endpoint actually expires silence
+		eval = ac.EvalAny(ac.EvalPermission(ac.ActionAlertingInstanceUpdate), ac.EvalPermission(ac.ActionAlertingSilencesWrite)) // delete endpoint actually expires silence
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/api/authorization.go:125]: Let's add read permissions here too.

**Review by yuri-tceretian — APPROVED** [review]
LGTM. Great work! I like how clean the service implementation is. Just a few non-blocking comments

File: pkg/services/ngalert/models/silence.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/grafana/alerting/notify"
+)
+
+// Silence is the model-layer representation of an alertmanager silence. Currently just a wrapper around the
+// alerting notify.Silence.
+type Silence notify.GettableSilence
+
+// GetRuleUID returns the rule UID of the silence if the silence is associated with a rule, otherwise nil.
+// Currently, this works by looking for a matcher with the RuleUIDLabel name and returning its value.
+func (s Silence) GetRuleUID() *string {
```
[the remark this answers is not in the document]
Comment (JacobsonMT) [reply pkg/services/ngalert/models/silence.go:16]: Tests are here: https://github.com/grafana/grafana/pull/86625/files#diff-62ce9cb0b47c6067d7fde1589a448ca6ac106762dc79eab9da388d72795996e1

File: pkg/tests/api/alerting/api_alertmanager_silence_test.go (hunk 3 above)
Comment (JacobsonMT) [reply pkg/tests/api/alerting/api_alertmanager_silence_test.go:108]: Ah, good catch. Forgot that detail, will update

File: pkg/services/ngalert/notifier/multiorg_alertmanager.go (hunk 1 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/notifier/multiorg_alertmanager.go:488]: Yes, we should probably rename it to something that makes it clearer in a separate PR.

File: pkg/services/ngalert/api/api_alertmanager_silences.go [hunk 7]
```diff
@@ -1,112 +1,76 @@
...
 	"github.com/grafana/grafana/pkg/util"
 )
 
-func (srv AlertmanagerSrv) RouteGetSilence(c *contextmodel.ReqContext, silenceID string) response.Response {
-	am, errResp := srv.AlertmanagerFor(c.SignedInUser.GetOrgID())
-	if errResp != nil {
-		return errResp
-	}
+// SilenceStore contains silence-specific methods for the RuleStore.
+type SilenceStore interface {
```
[the remark this answers is not in the document]
Comment (JacobsonMT) [reply pkg/services/ngalert/api/api_alertmanager_silences.go]: 13791634c1b398adf3bed1cb925bd98075f4009e

File: pkg/services/ngalert/api/persist.go [hunk 8]
```diff
@@ -26,4 +26,6 @@ type RuleStore interface {
 
 	// IncreaseVersionForAllRulesInNamespace Increases version for all rules that have specified namespace. Returns all rules that belong to the namespace
 	IncreaseVersionForAllRulesInNamespace(ctx context.Context, orgID int64, namespaceUID string) ([]ngmodels.AlertRuleKeyWithVersion, error)
+
+	SilenceStore
```
[the remark this answers is not in the document]
Comment (JacobsonMT) [reply pkg/services/ngalert/api/persist.go]: 13791634c1b398adf3bed1cb925bd98075f4009e

File: pkg/services/ngalert/notifier/silence_svc.go (hunk 2 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/notifier/silence_svc.go]: 27a08da83de23881ff47d1d918fd45f04139ced8

File: pkg/tests/api/alerting/api_alertmanager_silence_test.go (hunk 3 above)
Comment (JacobsonMT) [reply pkg/tests/api/alerting/api_alertmanager_silence_test.go]: b4131d342959fee41ed7e181d2e2314533b9f9a5

File: pkg/services/ngalert/api/authorization.go (hunk 4 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/api/authorization.go]: f8b72a89b7a16a08df43a3f6c95a93b01294016a

File: pkg/services/ngalert/api/authorization.go (hunk 5 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/api/authorization.go]: f8b72a89b7a16a08df43a3f6c95a93b01294016a

File: pkg/services/ngalert/models/silence.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuri-tceretian) [reply pkg/services/ngalert/models/silence.go:16]: Ha-ha.. That was from my previous review a week ago. Sorry!

File: pkg/services/ngalert/api/compat_silences.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package api
+
+import (
+	"github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+// Silence-specific compat functions to convert between API and model types.
+
+func GettableSilenceToAPIGettableSilence(s models.Silence) definitions.GettableSilence {
```
Comment (rwwiv) [inline pkg/services/ngalert/api/compat_silences.go:10]: These function names are very confusing, WDYT about `SilenceToGettableSilence` or something similar? The Postable/Gettable prefixes should probably stay solely for API models.

File: pkg/services/ngalert/models/testing.go [hunk 10]
```diff
@@ -903,3 +906,111 @@ func (n NotificationSettingsMutators) WithMuteTimeIntervals(muteTimeIntervals ..
 		ns.MuteTimeIntervals = muteTimeIntervals
 	}
 }
+
+// Silences
+
+// CopySilence creates a deep copy of Silence.
+func CopySilence(s Silence, mutators ...Mutator[Silence]) Silence {
```
Comment (rwwiv) [inline pkg/services/ngalert/models/testing.go:913]: Nit: WDYT about `CopySilenceWith`?

File: pkg/services/ngalert/store/alert_rule.go [hunk 11]
```diff
@@ -769,3 +770,20 @@ func ruleConstraintViolationToErr(rule ngmodels.AlertRule, err error) error {
 		return ngmodels.ErrAlertRuleConflict(rule, err)
 	}
 }
+
+// GetNamespacesByRuleUID returns a map of rule UIDs to their namespace UID.
+func (st DBstore) GetNamespacesByRuleUID(ctx context.Context, orgID int64, uids ...string) (map[string]string, error) {
```
Comment (rwwiv) [inline pkg/services/ngalert/store/alert_rule.go:775]: NB: Satisfying existing interface.

**Review by rwwiv — APPROVED** [review]
LGTM! Couple nits about confusing names, but this is looking very clean!

File: pkg/services/ngalert/api/compat_silences.go (hunk 9 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/api/compat_silences.go:10]: Ah good catch, these were named from an old version before I had the model struct and it shows :grimacing:

File: pkg/services/ngalert/models/testing.go (hunk 10 above)
Comment (JacobsonMT) [reply pkg/services/ngalert/models/testing.go]: 736a80fff25a0e6e7f346c72ddcba72dd22cfbe8

**Comment by JacobsonMT** [comment]
./deploy-to-hg

**Pull request merged**